components/script/dom/bindings/htmlconstructor.rs

```diff
@@ -343,3 +352,105 @@ pub fn pop_current_element_queue() {
     ScriptThread::push_new_element_queue();
 }
+
+unsafe fn prepare_html_constructor_prototype(
+    cx: JSContext,
+    args: CallArgs,
+    mut prototype: MutableHandleObject,
+    get_proto_object: fn(JSContext, HandleObject, MutableHandleObject),
+) -> Result<(), Error> {
+    // Step 2 https://html.spec.whatwg.org/multipage/#htmlconstructor
+    // The custom element definition cannot use an element interface as its constructor
+    // The new_target might be a cross-compartment wrapper. Get the underlying object
+    // so we can do the spec's object-identity checks.
+    rooted!(in(*cx) let new_target = UnwrapObjectDynamic(args.new_target().to_object(), *cx, 1));
+    if new_target.is_null() {
+        return Err(Error::Type("new.target is null".to_owned()));
+    }
+    if args.callee() == new_target.get() {
+        return Err(Error::Type(
+            "new.target must not be the active function object".to_owned(),
+        ));
+    }
+
+    // Step 6
+    rooted!(in(*cx) let mut proto_val = UndefinedValue());
+    let _ac = JSAutoRealm::new(*cx, new_target.get());
+    if !JS_GetProperty(
+        *cx,
+        new_target.handle(),
+        b"prototype\\0".as_ptr() as *const _,
+        proto_val.handle_mut(),
+    ) {
+        return Err(Error::JSFailed);
+    }
+    if !proto_val.is_object() {
+        // Step 7 of https://html.spec.whatwg.org/multipage/#htmlconstructor.
+        // This fallback behavior is designed to match analogous behavior for the
+        // JavaScript built-ins. So we enter the compartment of our underlying
+        // newTarget object and fall back to the prototype object from that global.
+        // XXX The spec says to use GetFunctionRealm(), which is not actually
+        // the same thing as what we have here (e.g. in the case of scripted callable proxies
+        // whose target is not same-compartment with the proxy, or bound functions, etc).
+        // https://bugzilla.mozilla.org/show_bug.cgi?id=1317658
+        rooted!(in(*cx) let global_object = CurrentGlobalOrNull(*cx));
+        get_proto_object(cx, global_object.handle(), prototype);
+    } else {
+        // Step 6
+        prototype.set(proto_val.to_object());
+    }
+    // Wrap prototype in this context since it is from the newTarget compartment
+    if !JS_WrapObject(*cx, RawMutableHandle::from(prototype)) {
+        return Err(Error::JSFailed);
+    }
+    return Ok(());
+}
+
+unsafe fn setup_html_constructor_prototype(
+    cx: JSContext,
+    reflector: HandleObject,
+    prototype: HandleObject,
+) -> bool {
+    rooted!(in(*cx) let mut element = reflector.get());
+    if !JS_WrapObject(*cx, RawMutableHandle::from(element.handle_mut())) {
+        return false;
+    }
+
+    JS_SetPrototype(
+        *cx,
+        RawHandle::from(element.handle()),
+        RawHandle::from(prototype),
+    );
+    return true;
+}
+
+pub unsafe fn call_html_constructor<T: DerivedFrom<Element> + DomObject>(
+    cx: JSContext,
+    args: CallArgs,
```

Comment:
Oh! I bet we need to make this `&mut CallArgs`, otherwise we're modifying a copy we've made of the structure and the one in the caller will be unchanged.